Reform proposal to add CrowdLevel

sccmcca opened the pull request:
Add a separate enum for CrowdLevel that does not conflict with OccupancyStatus.

aababilov committed:
Mention pathways.txt and attributions.txt in translations.txt

Fields in pathways.txt and attributions.txt  also may be translated.

paulswartz committed:
feat: `direction_id` no longer experimental (#210)

gcamp committed:
Add occupancy_percentage (#213)

gcamp committed:
Mark severity level as final (#214)

aababilov committed:
Move pathways, levels and attributions to the list of translatable tables

The following standard fields in those tables can be translated:

* pathways.txt - signposted_as, reversed_signposted_as
* levels.txt - level_name
* attributions.txt - organization_name, attribution_url, attribution_phone

Previously, the documentation placed them in a list of tables whose standard fields should not be translated.

barbeau committed:
fix: Add translations.txt to the files table (#218)

aababilov committed:
Update comment for pathways.signposted_as to allow translations

timMillet committed:
Add GTFS-ContinousStops (#208)

sccmcca committed:
Updated revision history for GTFS CHANGES.md (#225)

Updated GTFS revision history for transit/gtfs/CHANGES.md as of May 21, 2020.

barbeau committed:
feat: Add GitHub CI Action for GTFS-RT .proto validation  (#232)

barbeau committed:
feat: Label GTFS-RT ADDED trips as not fully specified (#230)

barbeau committed:
fix: Fix a typo in occupancy_percentage description (#234)

aababilov committed:
Fix a typo can can

barbeau committed:
feat: Alert `tts_header_text`, `tts_description_text` no longer experimental (#229)

barbeau committed:
Proposal: Support DUPLICATED trips in GTFS-RT  (#221)

* Support DUPLICATED trips in GTFS-RT

* Revert deprecation of ADDED

* Remove auto-format change

* Better explanation of "forbidden"

* Add experimental labels for message/fields

* Add experimental labels for field

* Add experimental label for message

* Clarify DUPLICATED use with VehiclePosition

* Clarify that true frequency-based trips can't be duplicated

* Explicitly define how the duplicated trip schedule and RT values are calculated

* docs: Add migration guide for transition from ADDED to DUPLICATED trips

* docs: Normalize comments in code

* docs: Allow trip.trip_id->trip_properties.trip_id link in migration guide

barbeau committed:
fix: Add extensions to TripProperties, remove extra comment (#238)

This is an editorial follow-up to https://github.com/google/transit/pull/221 to fix two items:
* New messages should have the extensions fields to allow 3rd party and internal extensions - this commit adds these fields
* An erroneous comment was added that labeled schedule_relationship as experimental - this commit removes it

lfontolliet committed:
Proposal: Support multi-car crowding in GTFS-RT (#237)

aababilov committed:
Change type of headway_secs to positive integer

headway_secs=0 does not have a physical meaning. This must be the time between departures for the same stop for a single trip.

barbeau committed:
Proposal: Support stop assignments in GTFS-RT (#219)

aababilov committed:
Define positive and non-null number field types (#251)

Define positive and non-null number field types

They are already used in the table descriptions.

Co-authored-by: scmcca <scott@mobilitydata.org>

timMillet committed:
Clarification of stop_times.stop_id (#258)

* Clarification of stop_times.stop_id

- It explicitly tells that the referenced locations must be stops/platforms and have a `stops.location_type` value of `0` or empty.
- Future-proof phrasing: it references only what must be done.

Commit:
Clarification of GTFS Realtime occupancy descriptions (#259)

Commit:
Add CrowdLevel

Commit:
style

barbeau commented on gtfs-realtime/proto/gtfs-realtime.proto:
```diff
@@ -503,11 +521,15 @@ message VehiclePosition {
     // -1 in case data is not available for this given carriage (as protobuf defaults to 0 otherwise)
     // This message/field is still experimental, and subject to change. It may be formally adopted in the future.
-    optional int32 occupancy_percentage = 4 [default = -1];
+    optional int32 occupancy_percentage = 5 [default = -1];
```
The .proto field value assignments are permanent and can't change after being added. So `occupancy_percentage` and `carriage_sequence` need to keep the same values, and `crowd_level` will need to have the value `6`.

barbeau commented on gtfs-realtime/spec/en/reference.md:
```diff
@@ -297,6 +301,20 @@ For describing passenger occupancy levels on a linear scale, see `occupancy_perc
 
+## _enum_ CrowdLevel
+
+The passenger crowd level for the vehicle or carriage. CrowdLevel enum values must be interpreted on an ordinal scale in increasing order from _**NOT_CROWDED**_ to _**CROWDED**_. 
```
I would add another sentence saying "In other words, producers must represent the minimum crowdedness level with _**NOT_CROWDED**_ and the maximum crowdedness level with _**CROWDED**_."

There are going to be questions as to when `CrowdLevel` should be used and when `OccupancyStatus` should be used. I'd be inclined to provide an explicit example to make sure there is no confusion.

barbeau replied:
Also, should producers take operational considerations like social distancing into account when publishing `CrowdLevel`? There is a line in `occupancy_percentage` definition about this - I'd suggest a similar mention here.

RTL-Gilles-Ferland replied:
As a producer, we prefer not to change the crowd level that we publish according to social distancing rules. We prefer to "tell it as it is" and let the client decide wether he is confortable or not with crowding in the vehicle. (since some clients are more nervous than others)

sccmcca replied:
Please see my [latest commit](https://github.com/TransitApp/transit/pull/3/commits/e96617fe799c907c869dc067a4f9efca6f6f7d0b) where I added the sentence "or as otherwise defined by current operating regulations". If not otherwise defined, I think there won't be any introduced ambiguity. Thoughts?

barbeau commented on gtfs-realtime/spec/en/reference.md:
```diff
@@ -297,6 +301,20 @@ For describing passenger occupancy levels on a linear scale, see `occupancy_perc
+| _**NOT_CROWDED**_ | _The vehicle or carriage is not considered crowded by most measures, having few or no passengers onboard._ |
+| _**SOME_CROWDING**_ | _The vehicle or carriage has some crowding by most measures, having some passengers onboard._ |
+| _**CROWDED**_ | _The vehicle or carriage is considered crowded by most measures, having many passengers onboard and limited space._ |
+| _**NO_DATA_AVAILABLE**_ | The vehicle of carriage does not have any crowd level data available at that time. |
```
Typo - `of` -> `or`

barbeau reviewed (commented):
Thanks @scmcca for putting this together. It looks pretty good to me. Some comments/thoughts in-line.

Commit:
Updates from review

- Reverted .proto field value assignments change
- Clarified ordinal usage of CrowdLevel
- Clarified CrowdLevel as subject to current allowed operating regulations (i.e., social distancing)
- Examples for simultaneous use of OccupancyStatus and CrowdLevel
- Corrected type

gcamp commented:
👍 on the actual field and description. 

🤷 Adding this on top of the two existing fields seems overkill, but I understand this is probably the solution that will get accepted vs replacement. 

👎 The ordering of the field changed. That makes the .proto not backward compatible.

sccmcca commented:
@gcamp Thank you for the review! 

I agree this proposal seems like overkill right now, but it's possible that it replaces the other experimental occupancy descriptions in the future. 

It looks like CrowdLevel should have been at value 11 under VehiclePosition, after occupancy_percentage that has value 10. I'll commit the fix!

Commit:
Fix value of CrowdLevel in VehiclePosition

Move to backward-compatible value

Commit:
Fix NO_DATA value

Commit:
Fix field number

Commit:
review

- Added missing default value
- Style and grammar
